File: LondonFhirService.Core/Services/Foundations/Providers/ProviderService.Exceptions.cs
// ---------------------------------------------------------
// Copyright (c) North East London ICB. All rights reserved.
// ---------------------------------------------------------

using System.Threading.Tasks;
using LondonFhirService.Core.Models.Foundations.Providers;
using LondonFhirService.Core.Models.Foundations.Providers.Exceptions;
using Xeptions;

namespace LondonFhirService.Core.Services.Foundations.Providers
{
    public partial class ProviderService
    {
        private delegate ValueTask<Provider> ReturningProviderFunction();

        private async ValueTask<Provider> TryCatch(ReturningProviderFunction returningProviderFunction)
        {
            try
            {
                return await returningProviderFunction();
            }
            catch (NullProviderServiceException nullProviderServiceException)
            {
                throw await CreateAndLogValidationException(nullProviderServiceException);
            }
            catch (InvalidProviderServiceException invalidProviderServiceException)
            {
                throw await CreateAndLogValidationException(invalidProviderServiceException);
            }
        }

        private async ValueTask<ProviderServiceValidationException> CreateAndLogValidationException(Xeption exception)
        {
            var providerServiceValidationException =
                new ProviderServiceValidationException(
                    message: "Provider validation errors occurred, please try again.",
                    innerException: exception);

            await this.loggingBroker.LogErrorAsync(providerServiceValidationException);

            return providerServiceValidationException;
        }
    }
}

File: LondonFhirService.Manage.Client/src/services/foundations/providers/providerService.exceptions.ts
import { ProviderApiBrokerException } from "../../../models/foundations/providers/exceptions/ProviderApiBrokerException";
import { ProviderDependencyException } from "../../../models/foundations/providers/exceptions/ProviderDependencyException";
import { ProviderServiceException } from "../../../models/foundations/providers/exceptions/ProviderServiceException";
import { ProviderValidationException } from "../../../models/foundations/providers/exceptions/ProviderValidationException";

export async function tryCatchProviderServiceAsync<T>(
    returningProviderFunction: () => Promise<T>)
    : Promise<T> {
    try {
        return await returningProviderFunction();
    } catch (exception) {
        if (exception instanceof ProviderValidationException) {
            throw exception;
        }

        if (exception instanceof ProviderApiBrokerException) {
            throw new ProviderDependencyException(
                "Provider dependency error occurred, please contact support.",
                exception);
        }

        throw new ProviderServiceException(
            "Provider service error occurred, please contact support.",
            exception);
    }
}
